File: analysis/mscalculate/cann/event_queue.py
# -------------------------------------------------------------------------
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is part of the MindStudio project.
#
# MindStudio is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#
#    http://license.coscl.org.cn/MulanPSL2
#
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# -------------------------------------------------------------------------

import logging

from mscalculate.cann.event import Event


class EventQueue:
    """
    This class is used for unified scheduling of events.
    1. Provides access to the event with the highest priority.
    """
    INIT_LEN = 100000

    def __init__(self, thread_id, init_len=INIT_LEN):
        self.thread_id = thread_id
        # for performance, use list to replace queue
        self.queue = [0] * init_len
        self.size = 0
        self.index = 0
        self._max_bound = -1
        self._lock = False

    def malloc_new_size(self, size: int):
        new_queue = [0] * size
        new_queue[: len(self.queue)] = self.queue
        self.queue = new_queue
        logging.info("malloc %d size for new thread %d queue", size, self.thread_id)

    def add(self, event: Event):
        if self._lock:
            logging.error("queue %d is locked, add function is illegal")
            return
        if self.size >= len(self.queue):
            self.malloc_new_size(2 * len(self.queue))
        self.queue[self.size] = event
        self._max_bound = max(event.bound, self._max_bound)
        self.size += 1

    def lock(self):
        self.queue[0:self.size] = sorted(self.queue[:self.size])
        self._lock = True

    def pop(self):
        if not self._lock:
            logging.error("queue %d is not locked, pop function is illegal")
            return Event.invalid_event()
        if self.index >= self.size:
            return Event.invalid_event()
        event = self.queue[self.index]
        self.index += 1
        return event

    def top(self) -> Event:
        if not self._lock:
            logging.error("queue %d is not locked, top function is illegal")
            return Event.invalid_event()
        if self.index >= self.size:
            return Event.invalid_event()
        return self.queue[self.index]

    def empty(self):
        return self.index == self.size

    def get_max_bound(self):
        return self._max_bound

File: analysis/csrc/domain/entities/tree/event_queue.cpp
/* -------------------------------------------------------------------------
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is part of the MindStudio project.
 *
 * MindStudio is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *
 *    http://license.coscl.org.cn/MulanPSL2
 *
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * -------------------------------------------------------------------------*/

#include <algorithm>

#include "analysis/csrc/infrastructure/dfx/log.h"
#include "analysis/csrc/infrastructure/utils/time_logger.h"
#include "analysis/csrc/domain/entities/tree/include/event_queue.h"

namespace Analysis {
namespace Domain {

EventQueue::EventQueue(uint32_t threadId, uint64_t initSize) : threadId_(threadId), initSize_(initSize)
{
    data_.reserve(initSize); // initSize = Api size + Tasktrack size
}

void EventQueue::Sort()
{
    std::string loggerName = "Sort event num " + std::to_string(tail_) +
        " in thread " + std::to_string(threadId_);
    Utils::TimeLogger logger(loggerName);
    auto comp = [](const std::shared_ptr<Event> &event1,
                   const std::shared_ptr<Event> &event2) {
        return event1->info.start < event2->info.start ||
            (event1->info.start == event2->info.start && event1->info.level > event2->info.level);
    };
    std::sort(data_.begin(), data_.begin() + tail_, comp);
}

void EventQueue::Push(const std::shared_ptr<Event> &event)
{
    if (!event) {
        ERROR("Event pointer is nullptr");
        return;
    }

    if (tail_ == initSize_) {
        ERROR("The number of added events exceeds the maximum EventQueue capacity");
        return;
    }
    data_.emplace_back(event);
    tail_ += 1;
    bound_ = std::max(event->info.end, bound_);
}

std::shared_ptr<Event> EventQueue::Pop()
{
    if (Empty()) {
        ERROR("Can not Pop when EventQueue is empty");
        return nullptr;
    }
    auto event = data_[head_];
    head_ += 1;
    return event;
}

std::shared_ptr<Event> EventQueue::Top()
{
    if (Empty()) {
        INFO("EventQueue for thread % is empty", threadId_);
        return nullptr;
    }
    return data_[head_];
}

bool EventQueue::Empty() const
{
    return head_ >= tail_;
}

uint64_t EventQueue::GetBound() const
{
    return bound_;
}

uint64_t EventQueue::GetSize() const
{
    if (Empty()) {
        return 0;
    }
    return tail_ - head_;
}

} // namespace Entities
} // namespace Analysis